File: finalResponse.js
import axios from "axios";

const OLLAMA_URL = process.env.OLLAMA_URL || "http://localhost:11434";

/**
 * Streams AI response token by token
 * @param {Array} chunks - Array of text/context chunks
 * @param {string} query - User question
 * @param {object} chatContext - Optional summary + messages
 * @param {function} onToken - Callback for each token
 */
export async function finalResponse(chunks, query, chatContext, onToken) {
  const contextText = chunks.map(c => c.text || c.raw_text || "").join("\n");
  const summaryText = chatContext.summary ? `Summary:\n${chatContext.summary}\n` : "";

  const prompt = `
You are a professional legal AI assistant. Use the context below to answer the user's question clearly and concisely.

Context:
${summaryText}${contextText}

User Question: ${query}
Answer in a narrative style, highlighting important decisions or info. Stream token by token.
  `;

  const response = await axios.post(
    `${OLLAMA_URL}/api/generate`,
    { model: "mistral", prompt, stream: true },
    { responseType: "stream" }
  );

  let buffer = "";
  for await (const chunk of response.data) {
    buffer += chunk.toString();
    const lines = buffer.split("\n").filter(Boolean);

    for (let line of lines) {
      line = line.replace(/^data:\s*/, "").trim();
      if (!line) continue;

      if (line === "[DONE]") {
        onToken("[DONE]");
        return;
      }

      try {
        const json = JSON.parse(line);
        if (json.response) onToken(json.response);
      } catch {
        buffer = line; // keep partial JSON for next chunk
      }
    }
    buffer = "";
  }

  onToken("[DONE]");
}

export default finalResponse;

File: services/final_response.py
# final_response.py
import asyncio
import httpx
import os
import json

OLLAMA_URL = os.getenv("OLLAMA_URL", "http://localhost:11434")


async def stream_final_response(chunks, query, chat_context, lang="en"):
    """
    Streams AI response token-by-token with bilingual support.
    lang: "en" for English, "pt" for Portuguese
    """
    context_text = "\n".join(c.get("text") or c.get("raw_text") or "" for c in chunks)
    summary_text = f"Summary:\n{chat_context.get('summary')}\n" if chat_context.get("summary") else ""

    # Map lang to instruction
    lang_instruction = {
        "en": "Answer the question in clear, concise English.",
        "pt": "Responda à pergunta de forma clara e concisa em Português."
    }.get(lang, "Answer clearly.")

    prompt = f"""
You are a professional legal AI assistant. {lang_instruction}

- Use the context below to answer the user's question clearly and concisely.
- If the context includes a URL, include it at the end of your answer in Markdown link format like: [View Law Here](https://example.com)
- Do NOT make up URLs. Only use those provided in context.
- Keep your response structured and human-readable.
- Highlight important legal details, articles, or clauses when relevant.

Context:
{summary_text}{context_text}

User Question: {query}

Answer in a narrative style.
Stream token by token.
"""

    async with httpx.AsyncClient(timeout=None) as client:
        async with client.stream(
            "POST", f"{OLLAMA_URL}/api/generate",
            json={"model": "mistral", "prompt": prompt, "stream": True}
        ) as resp:
            async for line in resp.aiter_lines():
                if not line:
                    continue
                if line.strip() == "[DONE]":
                    yield "[DONE]"
                    break

                # Clean up Ollama's `data:` lines
                clean = line.replace("data:", "").strip()
                try:
                    data = json.loads(clean)
                except Exception:
                    continue

                if "response" in data:
                    yield data["response"]
